plugins/spi/SPIOutput.cpp

```diff
@@ -909,31 +909,39 @@ void SPIOutput::IndividualWS2812bControl(const DmxBuffer &buffer) {
     return;
   }
 
-  const unsigned int length = std::min(m_pixel_count * WS2812B_SLOTS_PER_PIXEL,
-                                       buffer.Size() - first_slot);
+  const unsigned int length = m_pixel_count;
 
-  for (unsigned int i = 0; i < length / WS2812B_SLOTS_PER_PIXEL; i++) {
+  for (unsigned int i = 0; i < length; i++) {
     // Convert RGB to GRB
     unsigned int offset = first_slot + i * WS2812B_SLOTS_PER_PIXEL;
-    uint8_t r = buffer.Get(offset);
-    uint8_t g = buffer.Get(offset + 1);
-    uint8_t b = buffer.Get(offset + 2);
+
+    //Get DMX data
+    uint8_t r = 0;
+    uint8_t g = 0;
+    uint8_t b = 0;
+    if(offset < buffer.Size() - 2) {
+      r = buffer.Get(offset);
+      g = buffer.Get(offset + 1);
+      b = buffer.Get(offset + 2);
+    } // fill further pixel data only if the pixel data is empty
```

Comment:
I think this may stop it ever overwriting in future updates.

Reply:
Then test 3 should fail. This is where I prevent overreading the dmx buffer. The first time through it should fill the uninitialized pixels with 0 data, later passes it should only write data available from the dmx buffer. That is why it says else if, not just if.

The dmx standard specifies that a fixture loosing it's data input should keep doing what it was doing. I think this is the intended outcome.